File: python/src/nzdata/stats_nz.py
"""Client for the Aotearoa Data Explorer (ADE) API.

SDMX 2.1 REST, base URL https://api.data.stats.govt.nz/rest/. Server-side
only: SDMX responses can be large, so never call this from a browser.
"""

from __future__ import annotations

import csv
import io
import json
import math
import urllib.parse
import xml.etree.ElementTree as ET

import httpx
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional, Tuple

from .errors import StatsNzApiError, StatsNzError, StatsNzParseError

DEFAULT_BASE_URL = "https://api.data.stats.govt.nz/rest"
DEFAULT_VERSION = "1.0"
DEFAULT_TIMEOUT_MS = 30_000
USER_AGENT = "nz-open-data-connectors/0.1.0 (Language=Python)"
VALID_FORMATS = ("csv", "csvfilewithlabels", "jsondata")

FetchResult = Tuple[int, str]
FetchImpl = Callable[[str, Dict[str, str], int], FetchResult]


@dataclass
class StatsNzDataflow:
    """One table in the ADE catalogue."""

    id: str
    agency_id: str
    version: str
    title: str


@dataclass
class StatsNzObservation:
    """One row of data: dimension codes plus a value."""

    dimensions: Dict[str, str]
    value: Optional[float]
    labels: Optional[Dict[str, str]] = None
    status: Optional[str] = None


@dataclass
class StatsNzCodelistItem:
    """One dimension code mapped to a human label."""

    id: str
    name: str


@dataclass
class StatsNzCodelist:
    """A dimension codelist: codes to labels."""

    id: str
    agency_id: str
    version: str
    items: List[StatsNzCodelistItem]


def _default_fetch(url: str, headers: Dict[str, str], timeout_ms: int) -> FetchResult:
    try:
        response = httpx.get(
            url, headers=headers, timeout=timeout_ms / 1000, follow_redirects=True
        )
        return response.status_code, response.text
    except httpx.HTTPError as exc:
        raise StatsNzApiError(
            f"Stats NZ request failed: {exc}", status=0, retryable=True, url=url
        ) from exc


def parse_stats_nz_csv(text: str, dataflow_id: str) -> List[StatsNzObservation]:
    """Parses an ADE CSV payload (codes only) into typed observations."""
    reader = csv.DictReader(io.StringIO(text.lstrip("\ufeff")))
    if reader.fieldnames is None or "OBS_VALUE" not in reader.fieldnames:
        raise StatsNzParseError("Stats NZ CSV is missing the OBS_VALUE column")
    suffix = f"_{dataflow_id}"
    label_suffix = f"_LABEL{suffix}"
    observations: List[StatsNzObservation] = []
    for record in reader:
        dimensions: Dict[str, str] = {}
        labels: Dict[str, str] = {}
        for key, value in record.items():
            if key in ("DATAFLOW", "OBS_VALUE", "OBS_STATUS"):
                continue
            if key.endswith(label_suffix):
                labels[key[: -len(label_suffix)]] = value or ""
                continue
            if key.endswith(suffix):
                dimensions[key[: -len(suffix)]] = value or ""
        raw_value = record.get("OBS_VALUE") or ""
        value: Optional[float] = None if raw_value == "" else float(raw_value)
        if value is not None and not math.isfinite(value):
            raise StatsNzParseError(f"Invalid OBS_VALUE in Stats NZ CSV: {raw_value}")
        status = (record.get("OBS_STATUS") or "").strip()
        observation = StatsNzObservation(dimensions=dimensions, value=value)
        if labels:
            observation.labels = labels
        if status:
            observation.status = status
        observations.append(observation)
    return observations


def _format_value(value: Optional[float]) -> str:
    if value is None:
        return ""
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def _escape_csv_cell(value: str) -> str:
    if "," in value or '"' in value or "\n" in value:
        return f'"{value.replace(chr(34), chr(34) * 2)}"'
    return value


def serialize_stats_nz_rows_to_csv(rows: List[StatsNzObservation]) -> str:
    """Serializes typed observations back to CSV (dimensions, value, status)."""
    if not rows:
        return ""
    dimension_keys = sorted({key for row in rows for key in row.dimensions})
    has_status = any(row.status is not None for row in rows)
    header = dimension_keys + ["value"] + (["status"] if has_status else [])
    lines = [",".join(header)]
    for row in rows:
        cells = [row.dimensions.get(key, "") for key in dimension_keys]
        cells.append(_format_value(row.value))
        if has_status:
            cells.append(row.status or "")
        lines.append(",".join(_escape_csv_cell(cell) for cell in cells))
    return "\n".join(lines)


def _local_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def _as_record(value: Any) -> Optional[Dict[str, Any]]:
    return value if isinstance(value, dict) else None


def _as_string(value: Any) -> str:
    return value if isinstance(value, str) else ""


def _simplify(element: ET.Element) -> Dict[str, Any]:
    """Converts an ElementTree element into plain dicts keyed by local names."""
    result: Dict[str, Any] = {}
    for key, value in element.attrib.items():
        result[f"@_{_local_name(key)}"] = value
    children = list(element)
    if not children:
        if element.text and element.text.strip():
            result["#text"] = element.text.strip()
        return result
    for child in children:
        name = _local_name(child.tag)
        child_value = _simplify(child)
        if name in result:
            if not isinstance(result[name], list):
                result[name] = [result[name]]
            result[name].append(child_value)
        else:
            result[name] = child_value
    return result


def parse_dataflow_catalogue_xml(xml: str) -> List[StatsNzDataflow]:
    """Parses the ADE dataflow catalogue XML into dataflow records."""
    try:
        root = ET.fromstring(xml)
    except ET.ParseError as exc:
        raise StatsNzParseError("Failed to parse Stats NZ dataflow catalogue XML") from exc
    doc = {_local_name(root.tag): _simplify(root)}
    structure = _as_record(_as_record(doc).get("Structure"))
    if structure is None:
        raise StatsNzParseError("Stats NZ dataflow catalogue XML has no Structure element")
    structures = _as_record(structure.get("Structures"))
    dataflows_node = _as_record(structures.get("Dataflows")) if structures else None
    dataflows_value = dataflows_node.get("Dataflow") if dataflows_node else None
    dataflows = (
        dataflows_value
        if isinstance(dataflows_value, list)
        else [dataflows_value]
        if dataflows_value is not None
        else []
    )
    return [
        StatsNzDataflow(
            id=_as_string(_as_record(flow).get("@_id")),
            agency_id=_as_string(_as_record(flow).get("@_agencyID")),
            version=_as_string(_as_record(flow).get("@_version")),
            title=_as_string(_as_record(_as_record(flow).get("Name")).get("#text")),
        )
        for flow in dataflows
    ]


def _to_codelist_items(codes_value: Any) -> List[StatsNzCodelistItem]:
    codes = (
        codes_value
        if isinstance(codes_value, list)
        else [codes_value]
        if codes_value is not None
        else []
    )
    return [
        StatsNzCodelistItem(
            id=_as_string(_as_record(code).get("@_id")),
            name=_as_string(_as_record(_as_record(code).get("Name")).get("#text")),
        )
        for code in codes
    ]


def parse_codelist_xml(xml: str) -> StatsNzCodelist:
    """Parses an ADE codelist XML payload into a codelist."""
    try:
        root = ET.fromstring(xml)
    except ET.ParseError as exc:
        raise StatsNzParseError("Failed to parse Stats NZ codelist XML") from exc
    doc = {_local_name(root.tag): _simplify(root)}
    structure = _as_record(_as_record(doc).get("Structure"))
    structures = _as_record(structure.get("Structures")) if structure else None
    codelists = _as_record(structures.get("Codelists")) if structures else None
    codelists_value = codelists.get("Codelist") if codelists else None
    codelist_list = (
        codelists_value
        if isinstance(codelists_value, list)
        else [codelists_value]
        if codelists_value is not None
        else []
    )
    codelist = _as_record(codelist_list[0]) if codelist_list else None
    if codelist is None:
        raise StatsNzParseError("Stats NZ codelist XML has no Codelist element")
    return StatsNzCodelist(
        id=_as_string(codelist.get("@_id")),
        agency_id=_as_string(codelist.get("@_agencyID")),
        version=_as_string(codelist.get("@_version")),
        items=_to_codelist_items(codelist.get("Code")),
    )


class StatsNzClient:
    """Typed client for the Aotearoa Data Explorer API."""

    def __init__(
        self,
        subscription_key: Optional[str] = None,
        base_url: str = DEFAULT_BASE_URL,
        timeout_ms: int = DEFAULT_TIMEOUT_MS,
        fetch_impl: Optional[FetchImpl] = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._subscription_key = subscription_key
        self._timeout_ms = timeout_ms
        self._fetch = fetch_impl or _default_fetch

    def _send_request(self, path: str, accept: str) -> str:
        url = f"{self._base_url}{path}"
        headers: Dict[str, str] = {
            "Accept": accept,
            "Cache-Control": "no-cache",
            "user-agent": USER_AGENT,
        }
        if self._subscription_key is not None:
            headers["Ocp-Apim-Subscription-Key"] = self._subscription_key
        status, body = self._fetch(url, headers, self._timeout_ms)
        if status < 200 or status >= 300:
            message = f"Stats NZ API request failed with status {status}"
            try:
                parsed = json.loads(body)
                if isinstance(parsed, dict) and isinstance(parsed.get("message"), str):
                    message = parsed["message"]
            except json.JSONDecodeError:
                pass
            raise StatsNzApiError(
                message, status=status, retryable=status == 429 or status >= 500, url=url
            )
        return body

    def get_data(
        self,
        dataflow_id: str,
        key: str = "all",
        version: Optional[str] = None,
        format: str = "csv",
    ) -> List[StatsNzObservation]:
        """Pulls data rows for a dataflow. Keyless for AGR_* tables."""
        if format not in VALID_FORMATS:
            raise StatsNzError(f"Unsupported Stats NZ response format: {format}")
        if format == "jsondata":
            raise StatsNzError("jsondata format is not implemented in the Python port")
        version = version or DEFAULT_VERSION
        path = (
            f"/data/STATSNZ,{urllib.parse.quote(dataflow_id)},{urllib.parse.quote(version)}"
            f"/{urllib.parse.quote(key)}?format={format}"
        )
        body = self._send_request(path, "text/csv")
        return parse_stats_nz_csv(body, dataflow_id)

    def get_dataflow_catalogue(self) -> List[StatsNzDataflow]:
        """Lists every dataflow in the ADE catalogue. Keyless."""
        body = self._send_request("/dataflow/STATSNZ/all", "application/xml")
        return parse_dataflow_catalogue_xml(body)

    def get_codelist(
        self, codelist_id: str, version: Optional[str] = None
    ) -> StatsNzCodelist:
        """Resolves dimension codes to labels. Needs a subscription key."""
        version_segment = "" if version is None else f"/{urllib.parse.quote(version)}"
        path = f"/codelist/STATSNZ/{urllib.parse.quote(codelist_id)}{version_segment}"
        body = self._send_request(path, "application/xml")
        return parse_codelist_xml(body)


def create_stats_nz_client(
    subscription_key: Optional[str] = None,
    base_url: str = DEFAULT_BASE_URL,
    timeout_ms: int = DEFAULT_TIMEOUT_MS,
    fetch_impl: Optional[FetchImpl] = None,
) -> StatsNzClient:
    """Builds a Stats NZ client. Keys are read from the caller, never the network."""
    return StatsNzClient(
        subscription_key=subscription_key,
        base_url=base_url,
        timeout_ms=timeout_ms,
        fetch_impl=fetch_impl,
    )

File: ruby/lib/nzdata/stats_nz.rb
# frozen_string_literal: true

require 'cgi'
require 'csv'
require 'json'
require 'net/http'
require 'rexml/document'
require 'uri'

module Nzdata
  StatsNzDataflow = Struct.new(:id, :agency_id, :version, :title)
  StatsNzObservation = Struct.new(:dimensions, :value, :labels, :status)
  StatsNzCodelistItem = Struct.new(:id, :name)
  StatsNzCodelist = Struct.new(:id, :agency_id, :version, :items)

  DEFAULT_BASE_URL = 'https://api.data.stats.govt.nz/rest'
  DEFAULT_VERSION = '1.0'
  DEFAULT_TIMEOUT_MS = 30_000
  USER_AGENT = 'nz-open-data-connectors/0.1.0 (Language=Ruby)'
  VALID_FORMATS = %w[csv csvfilewithlabels jsondata].freeze

  module_function

  def parse_stats_nz_csv(text, dataflow_id)
    text = text.sub(/\A\uFEFF/, '')
    rows = CSV.parse(text, headers: true)
    return [] if rows.nil? || rows.empty?

    headers = rows.headers
    unless headers.include?('OBS_VALUE')
      raise StatsNzParseError,
            'Stats NZ CSV is missing the OBS_VALUE column'
    end

    suffix = "_#{dataflow_id}"
    label_suffix = "_LABEL#{suffix}"
    rows.map do |row|
      dimensions = {}
      labels = {}
      row.to_h.each do |key, value|
        next if %w[DATAFLOW OBS_VALUE OBS_STATUS].include?(key)

        if key.end_with?(label_suffix)
          labels[key[0...-label_suffix.length]] = value.to_s
          next
        end
        dimensions[key[0...-suffix.length]] = value.to_s if key.end_with?(suffix)
      end
      raw_value = row['OBS_VALUE'].to_s
      value = raw_value.empty? ? nil : Float(raw_value)
      if value && !value.finite?
        raise StatsNzParseError, "Invalid OBS_VALUE in Stats NZ CSV: #{raw_value}"
      end

      observation = StatsNzObservation.new(dimensions, value)
      observation.labels = labels unless labels.empty?
      observation.status = row['OBS_STATUS'].to_s.strip unless row['OBS_STATUS'].to_s.strip.empty?
      observation
    end
  end

  def format_value(value)
    return '' if value.nil?

    value == value.to_i ? value.to_i.to_s : value.to_s
  end

  def escape_csv_cell(value)
    if value.include?(',') || value.include?('"') || value.include?("\n")
      "\"#{value.gsub('"', '""')}\""
    else
      value
    end
  end

  def serialize_stats_nz_rows_to_csv(rows)
    return '' if rows.empty?

    dimension_keys = rows.flat_map { |row| row.dimensions.keys }.uniq.sort
    has_status = rows.any? { |row| !row.status.nil? }
    header = dimension_keys + ['value'] + (has_status ? ['status'] : [])
    lines = [header.join(',')]
    rows.each do |row|
      cells = dimension_keys.map { |key| row.dimensions[key].to_s }
      cells << format_value(row.value)
      cells << row.status.to_s if has_status
      lines << cells.map { |cell| escape_csv_cell(cell) }.join(',')
    end
    lines.join("\n")
  end

  def parse_dataflow_catalogue_xml(xml)
    doc = REXML::Document.new(xml)
    structure = REXML::XPath.first(doc, '//*[local-name()="Structure"]')
    if structure.nil?
      raise StatsNzParseError,
            'Stats NZ dataflow catalogue XML has no Structure element'
    end

    REXML::XPath.match(doc, '//*[local-name()="Dataflow"]').map do |flow|
      name = REXML::XPath.first(flow, './/*[local-name()="Name"]')
      StatsNzDataflow.new(
        flow.attributes['id'].to_s,
        flow.attributes['agencyID'].to_s,
        flow.attributes['version'].to_s,
        name ? name.text.to_s : ''
      )
    end
  rescue REXML::ParseException
    raise StatsNzParseError, 'Failed to parse Stats NZ dataflow catalogue XML'
  end

  def parse_codelist_xml(xml)
    doc = REXML::Document.new(xml)
    codelist = REXML::XPath.first(doc, '//*[local-name()="Codelist"]')
    raise StatsNzParseError, 'Stats NZ codelist XML has no Codelist element' if codelist.nil?

    items = REXML::XPath.match(codelist, './/*[local-name()="Code"]').map do |code|
      name = REXML::XPath.first(code, './/*[local-name()="Name"]')
      StatsNzCodelistItem.new(code.attributes['id'].to_s, name ? name.text.to_s : '')
    end
    StatsNzCodelist.new(
      codelist.attributes['id'].to_s,
      codelist.attributes['agencyID'].to_s,
      codelist.attributes['version'].to_s,
      items
    )
  rescue REXML::ParseException
    raise StatsNzParseError, 'Failed to parse Stats NZ codelist XML'
  end

  class StatsNzClient
    def initialize(subscription_key: nil, base_url: DEFAULT_BASE_URL,
                   timeout_ms: DEFAULT_TIMEOUT_MS, fetch_impl: nil)
      @base_url = base_url.sub(%r{/+$}, '')
      @subscription_key = subscription_key
      @timeout_ms = timeout_ms
      @fetch = fetch_impl || method(:default_fetch)
    end

    def get_data(dataflow_id, key: 'all', version: nil, format: 'csv')
      unless VALID_FORMATS.include?(format)
        raise StatsNzError,
              "Unsupported Stats NZ response format: #{format}"
      end
      if format == 'jsondata'
        raise StatsNzError,
              'jsondata format is not implemented in the Ruby port'
      end

      version ||= DEFAULT_VERSION
      path = "/data/STATSNZ,#{encode(dataflow_id)},#{encode(version)}/" \
             "#{encode(key)}?format=#{format}"
      body = send_request(path, 'text/csv')
      Nzdata.parse_stats_nz_csv(body, dataflow_id)
    end

    def get_dataflow_catalogue
      body = send_request('/dataflow/STATSNZ/all', 'application/xml')
      Nzdata.parse_dataflow_catalogue_xml(body)
    end

    def get_codelist(codelist_id, version: nil)
      version_segment = version.nil? ? '' : "/#{encode(version)}"
      path = "/codelist/STATSNZ/#{encode(codelist_id)}#{version_segment}"
      body = send_request(path, 'application/xml')
      Nzdata.parse_codelist_xml(body)
    end

    private

    def encode(value)
      URI.encode_www_form_component(value)
    end

    def send_request(path, accept)
      url = "#{@base_url}#{path}"
      headers = { 'Accept' => accept, 'Cache-Control' => 'no-cache', 'user-agent' => USER_AGENT }
      headers['Ocp-Apim-Subscription-Key'] = @subscription_key unless @subscription_key.nil?
      status, body = @fetch.call(url, headers, @timeout_ms)
      unless status.between?(200, 299)
        message = "Stats NZ API request failed with status #{status}"
        begin
          parsed = JSON.parse(body)
          message = parsed['message'] if parsed.is_a?(Hash) && parsed['message'].is_a?(String)
        rescue JSON::ParserError
          # keep the default message
        end
        raise StatsNzApiError.new(message, status: status,
                                           retryable: status == 429 || status >= 500, url: url)
      end
      body
    end

    def default_fetch(url, headers, timeout_ms)
      uri = URI(url)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == 'https'
      http.open_timeout = timeout_ms / 1000.0
      http.read_timeout = timeout_ms / 1000.0
      request = Net::HTTP::Get.new(uri)
      headers.each { |key, value| request[key] = value }
      response = http.request(request)
      [response.code.to_i, response.body.to_s]
    rescue StandardError => e
      raise StatsNzApiError.new("Stats NZ request failed: #{e.message}", status: 0,
                                                                         retryable: true, url: url)
    end
  end

  def create_stats_nz_client(subscription_key: nil, base_url: DEFAULT_BASE_URL,
                             timeout_ms: DEFAULT_TIMEOUT_MS, fetch_impl: nil)
    StatsNzClient.new(
      subscription_key: subscription_key,
      base_url: base_url,
      timeout_ms: timeout_ms,
      fetch_impl: fetch_impl
    )
  end
end
